File: services/core/src/main/java/me/unbrdn/core/interview/application/port/in/GetInterviewHistoryUseCase.java
package me.unbrdn.core.interview.application.port.in;

import java.util.List;

/** Use Case for retrieving interview history */
public interface GetInterviewHistoryUseCase {

    /**
     * Get interview message history
     *
     * @param interviewId the interview ID
     * @return list of interview messages
     */
    List<InterviewMessageDto> execute(String interviewId);

    /** DTO for interview message */
    record InterviewMessageDto(
            String role,
            String type,
            String content,
            String timestamp,
            java.util.Map<String, String> payload) {}
}

File: services/gateways/bff/src/modules/interview/usecases/get-interview-history.usecase.ts
import { Injectable } from "@nestjs/common";
import { InterviewGrpcService } from "../../../infra/grpc/services/interview-grpc.service";

export class GetInterviewHistoryCommand {
    constructor(public readonly interviewId: string) {}
}

@Injectable()
export class GetInterviewHistoryUseCase {
    constructor(private readonly interviewGrpcService: InterviewGrpcService) {}

    async execute(command: GetInterviewHistoryCommand) {
        const response = await this.interviewGrpcService.getHistory(command.interviewId);

        return response.messages.map((m: any) => ({
            role: m.role,
            type: m.type,
            content: m.content,
            timestamp: m.timestamp,
            payload: m.payload,
        }));
    }
}
